python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py

```diff
@@ -472,6 +475,16 @@ def stream_raw_events(self) -> Iterator[DbtCliEventMessage]:
                 if is_error_message:
                     self._error_messages.append(str(event))
 
+                # Attempt to parse the columns metadata from the event message.
+                # If it exists, then save it as metadata for the dbt node in the manifest.
+                if event.raw_event["info"]["name"] == "JinjaLogInfo":
+                    with contextlib.suppress(orjson.JSONDecodeError):
+                        columns = orjson.loads(event.raw_event["info"]["msg"])
+                        metadata_by_unique_id[cast(str, unique_id)] = {"columns": columns}
```

Comment:
Is my understanding correct that this line can mutate the metadata dict that's on other `DbtCliEventMessage`s than the one being processed in this iteration of the loop?  I think it would be unexpected, from the perspective of the consumer of `stream_raw_events`, for events that come out of it to be mutated after the fact.

Would it make sense to move some of this logic up a level into `stream`? E.g. something like (didn't think too deeply about the names):
```python
        additional_metadata_by_unique_id = {}
        for event in self.stream_raw_events():
            if event.contains_additional_metadata():
                additional_metadata_by_unique_id[event.unique_id] = event.additional_metadata

            yield from event.to_default_asset_events(
                manifest=self.manifest,
                dagster_dbt_translator=self.dagster_dbt_translator,
                context=self.context,
                additional_metadata=additional_metadata_by_unique_id[event.unique_id]
            )
```

This would confine any mutation to within a function call.

Reply:
> Is my understanding correct that this line can mutate the metadata dict that's on other DbtCliEventMessages than the one being processed in this iteration of the loop?

We only processed the `NodeFinished` event when new call `to_default_asset_events`, which is the final event for a given node. At that point, the metadata will have been set and cannot be mutated after the fact.

I am happy to refactor this so that we pass in a copy of the metadata so that we defend ourselves against metadata changing underneath our feet, but I don't believe that will happen when we process the dbt event stream.

> Would it make sense to move some of this logic up a level into stream?

I don't think this is necessary if we just copy the metadata. The metadata will still be attached to the event in this case, so no need to add it as an argument to `to_default_asset_events`.